internal/datacoord/segment_reference_manager.go

```diff
@@ -0,0 +1,199 @@
+	}
+	keys, _, err := segReferManager.etcdKV.LoadWithPrefix(segmentReferPrefix)
+	if err != nil {
+		log.Error("load segments lock from etcd failed", zap.Error(err))
+		return nil, err
+	}
+
+	for _, key := range keys {
+		segID, nodeID, err := parseLockKey(key)
+		if err != nil {
+			log.Error("parse segment lock key failed", zap.String("lock key", key), zap.Error(err))
+			return nil, err
+		}
+		segLock := &SegmentLock{
+			segmentID: segID,
+			nodeID:    nodeID,
+			locKey:    key,
+		}
+		segReferManager.segmentsLock[segID] = append(segReferManager.segmentsLock[segID], segLock)
+	}
+	return segReferManager, nil
+}
+
+func (srm *SegmentReference) AddSegmentsLock(segIDs []UniqueID, nodeID UniqueID) error {
+	srm.lock.Lock()
+	defer srm.lock.Unlock()
+
+	locKVs := make(map[string]string)
+	segID2SegmentLock := make(map[UniqueID][]*SegmentLock)
+	for _, segID := range segIDs {
+		locKey := path.Join(segmentReferPrefix, strconv.FormatInt(nodeID, 10), strconv.FormatInt(segID, 10))
+		locKVs[locKey] = strconv.FormatInt(nodeID, 10)
+		segLock := &SegmentLock{
+			segmentID: segID,
+			nodeID:    nodeID,
+			locKey:    locKey,
+		}
+		segID2SegmentLock[segID] = append(segID2SegmentLock[segID], segLock)
+	}
+	if err := srm.etcdKV.MultiSave(locKVs); err != nil {
+		log.Error("AddSegmentsLock save  segment lock to etcd failed", zap.Int64s("segIDs", segIDs), zap.Error(err))
+		return err
+	}
+	for segID, segLocks := range segID2SegmentLock {
+		srm.segmentsLock[segID] = append(srm.segmentsLock[segID], segLocks...)
+	}
+	return nil
+}
+
+func (srm *SegmentReference) ReleaseSegmentsLock(segIDs []UniqueID, nodeID UniqueID) error {
+	srm.lock.Lock()
+	defer srm.lock.Unlock()
+
+	locKeys := make([]string, 0)
+	for _, segID := range segIDs {
+		if _, ok := srm.segmentsLock[segID]; !ok {
+			continue
+		}
+		for i := 0; i < len(srm.segmentsLock[segID]); i++ {
+			segLock := srm.segmentsLock[segID][i]
+			if segLock.nodeID == nodeID {
+				locKeys = append(locKeys, segLock.locKey)
+				srm.segmentsLock[segID] = append(srm.segmentsLock[segID][:i], srm.segmentsLock[segID][i+1:]...)
+				i--
+			}
+		}
+		if len(srm.segmentsLock[segID]) == 0 {
+			delete(srm.segmentsLock, segID)
+		}
+	}
+	return srm.etcdKV.MultiRemove(locKeys)
+}
+
+func (srm *SegmentReference) ReleaseSegmentsLockByNodeID(nodeID UniqueID) error {
+	srm.lock.Lock()
+	defer srm.lock.Unlock()
+
+	locKeys := make([]string, 0)
+	for segID := range srm.segmentsLock {
+		for i := 0; i < len(srm.segmentsLock[segID]); i++ {
+			segLock := srm.segmentsLock[segID][i]
+			if segLock.nodeID == nodeID {
+				locKeys = append(locKeys, segLock.locKey)
+				srm.segmentsLock[segID] = append(srm.segmentsLock[segID][:i], srm.segmentsLock[segID][i+1:]...)
+				i--
+			}
+		}
+		if len(srm.segmentsLock[segID]) == 0 {
+			delete(srm.segmentsLock, segID)
+		}
+	}
+	return srm.etcdKV.MultiRemove(locKeys)
+}
+
+func (srm *SegmentReference) ReleaseSegmentLockByOnlineIDs(nodeIDs []UniqueID) error {
```

Comment:
why we need this call?

Reply:
I get what you are doing. 
The naming is a little confusing,  I thought this should happened same as NewSegmentReference